File: Chatroom_java/app/src/main/java/com/example/chatroom_java/MainActivity.java
package com.example.chatroom_java;

import static android.Manifest.permission.RECORD_AUDIO;
import static com.example.chatroom_java.data.LoadData.loadJSONFromAsset;
import static com.example.chatroom_java.data.LoadData.parseChatJSON;
import static com.example.chatroom_java.data.SocketClient.*;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.media.MediaRecorder;
import android.os.Bundle;
import android.os.Environment;
import android.os.SystemClock;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;
import android.os.AsyncTask;


import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.loader.content.AsyncTaskLoader;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.chatroom_java.chatList.ChatAdapter;
import com.example.chatroom_java.data.Chat;
import com.example.chatroom_java.data.DataSource;
import com.example.chatroom_java.data.SocketClient;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.io.BufferedOutputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.Socket;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ChatAdapter adapter;
    private boolean isRecording= false;
    private Button recButton;
    private MediaRecorder mediaRecorder;
    private static String recordFile = null;
    private static String recordFilePath = null;
    public static final int REQUEST_AUDIO_PERMISSION_CODE = 1;
    private DataSource dataSource;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        //別動-------------------------------------------------------------------------------
        Log.d("Wellcome","Wellcome");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recycler_view);

        dataSource = DataSource.getDataSource(getResources());
        final java.util.List<Chat> chatList = dataSource.getChatList().getValue();

        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        adapter = new ChatAdapter(this, chatList != null ? chatList : new ArrayList<Chat>(), recyclerView);
        recyclerView.setAdapter(adapter);


        recButton = findViewById(R.id.rec_button);
    }
    //別動--------------------------------------------------------------------------------------------
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        switch (requestCode) {
            case REQUEST_AUDIO_PERMISSION_CODE:
                if (grantResults.length >= 2) { // Check if there are at least two results
                    boolean permissionToRecord = grantResults[0] == PackageManager.PERMISSION_GRANTED;
                    boolean permissionToStore = grantResults[1] == PackageManager.PERMISSION_GRANTED;
                    if (permissionToRecord && permissionToStore) {
                        Toast.makeText(getApplicationContext(), "Permission Granted", Toast.LENGTH_LONG).show();
                        onRestart();
                    } else {
                        Toast.makeText(getApplicationContext(), "Permission Denied", Toast.LENGTH_LONG).show();
                    }
                }
                break;
        }
    }
    public boolean CheckPermissions() {
        int result1 = ContextCompat.checkSelfPermission(getApplicationContext(), RECORD_AUDIO);
        return result1 == PackageManager.PERMISSION_GRANTED;
    }
    private void RequestPermissions() {
        ActivityCompat.requestPermissions(MainActivity.this, new String[]{RECORD_AUDIO}, REQUEST_AUDIO_PERMISSION_CODE);
    }

    public void recording(View view) throws IOException {
        if (CheckPermissions()) {
            if(isRecording){
                stopRecording();
                isRecording = false;

            }
            else{
                startRecording();
                isRecording = true;
            }
        } else {
            RequestPermissions();
        }

        String json = loadJSONFromAsset(getApplicationContext(), "chats.json");
        List<Chat> chatList = parseChatJSON(json);
        List<Chat> currentChatList = new ArrayList<>(dataSource.getChatList().getValue());
        currentChatList.addAll(chatList);

        dataSource.getChatList().setValue(currentChatList);
        adapter.updateData(currentChatList);

        recyclerView.postDelayed(new Runnable() {
            @Override
            public void run() {
                recyclerView.smoothScrollToPosition(currentChatList.size() - 1);
            }
        }, 100);
    }

    private void startRecording() throws IOException {

        recButton.setBackgroundResource(R.drawable.square_shape);
        recordFilePath = this.getExternalFilesDir("/").getAbsolutePath();
        Log.d("recordFilePath",recordFilePath);
        SimpleDateFormat format1 = new SimpleDateFormat("yyyy_MM_dd_hh_mm_ss", Locale.ENGLISH);
        Date current = new Date();
        recordFile = "Recording_"+ format1.format(current) +".wav";
        Log.d("recordFilePath",recordFilePath);
        Log.d("recordFile",recordFile);
        Log.d("recordFileSize", String.valueOf(recordFile.length()));
        Toast toastStart = Toast.makeText(this, "Recording started",Toast.LENGTH_SHORT);
        toastStart.show();
        mediaRecorder = new MediaRecorder();
        mediaRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
        mediaRecorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
        mediaRecorder.setOutputFile(recordFilePath + "/" + recordFile);
        mediaRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
        try {
            mediaRecorder.prepare();
        } catch (IOException e) {
            e.printStackTrace();
        }
        mediaRecorder.start();

    }


    private void stopRecording() {
        if (mediaRecorder != null) {
            try {
                mediaRecorder.stop();
                Log.d("ZFinish","finish");
            } catch (IllegalStateException e) {
                e.printStackTrace();
            }

            mediaRecorder.reset();
            mediaRecorder.release();
            mediaRecorder = null;

            String serverAddress = "172.16.168.1";
            int serverPort = 8082;
            new SocketClient(serverAddress, serverPort, recordFilePath + "/" + recordFile).execute();

            recButton.setBackgroundResource(R.drawable.record_button_background);
            Toast toastStop = Toast.makeText(getBaseContext(), "Recording stopped", Toast.LENGTH_SHORT);
            toastStop.show();
        }
    }
    @Override
    public void onStop() {
        super.onStop();
        if (isRecording) {
            stopRecording();
        }
    }
}

File: chatroom/app/src/main/java/com/example/chatroom_wav/MainActivity.kt
package com.example.chatroom_wav

import ChatAdapter
import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.Button
import android.widget.ImageView
import android.widget.Toast
import android.widget.Toolbar
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.chatroom_wav.data.Chat
import com.example.chatroom_wav.data.DataSource
import com.example.chatroom_wav.data.SocketClient
import com.example.chatroom_wav.data.loadJSONFromAsset
import com.example.chatroom_wav.data.parseChatJSON
import com.example.chatroom_wav.setting.SettingActivity
import com.example.chatroom_wav.wave.RecorderState
import com.example.chatroom_wav.wave.WaveRecorder
import java.util.Timer
import java.util.TimerTask



class MainActivity : AppCompatActivity() {


    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: ChatAdapter

    private val PERMISSIONS_REQUEST_RECORD_AUDIO = 77

    private lateinit var waveRecorder: WaveRecorder
    private lateinit var filePath: String
    private var isRecording = false

    private var timer: Timer? = null
    private var isTimerScheduled = false

    private var serverAddress :String = "172.17.40.106"
    private var serverPort :Int = 8083
    private var language :String = "en"


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //直接讀檔顯示頭兩個chat(之後刪掉)======================================================
        recyclerView = findViewById(R.id.recycler_view)

        val dataSource = DataSource.getDataSource(resources)
        val chatList = dataSource.getChatList().value?.toMutableList() ?: mutableListOf()

        recyclerView.layoutManager =
            LinearLayoutManager(this)

        adapter = ChatAdapter(this, chatList ?: emptyList() ,recyclerView)
        recyclerView.adapter = adapter

        //setting========================================================================
        val settingIcon = findViewById<ImageView>(R.id.setting_icon)

        settingIcon.setOnClickListener {
            val intent = Intent(this, SettingActivity::class.java)
            intent.putExtra("serverAddress", serverAddress)
            intent.putExtra("serverPort", serverPort)
            intent.putExtra("language", language)
            startActivity(intent)
        }
        val sharedPreferences = getSharedPreferences("MyPrefs", Context.MODE_PRIVATE)
        serverAddress = sharedPreferences.getString("serverAddress", "172.17.40.106") ?: "172.17.40.106"
        serverPort = sharedPreferences.getString("serverPort", "8083")?.toIntOrNull() ?: 8083
        language = sharedPreferences.getString("language", "en") ?: "en"
        Log.d("MainActivity", "Server Address: $serverAddress, Server Port: $serverPort")

        //開權限===========================================================================
        if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.RECORD_AUDIO
            )
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                PERMISSIONS_REQUEST_RECORD_AUDIO
            )
        }

        //設定音訊存檔路徑===================================================================
        val downloadDir =
            Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS)

        if (!downloadDir.exists()) {
            downloadDir.mkdirs()
        }
        //val recordFilePath = downloadDir.absolutePath
        val recordFilePath = this.getCacheDir().getAbsolutePath()
        Log.d("recordFilePath",recordFilePath)
        //initial filepath to 1.wav
        var isWav2 = false
        filePath = recordFilePath + "/audioFile1.wav"

        //設定錄音工具=====================================================================
        waveRecorder = WaveRecorder(filePath)

        waveRecorder.onStateChangeListener = {
            when (it) {
                RecorderState.RECORDING -> startRecording()
                RecorderState.STOP -> stopRecording()
                else -> {}
            }
        }
        waveRecorder.noiseSuppressorActive=true

        //每三秒做一次顯示========================================================================
        var count = 0
        var transcript: String = ""
        var filepath_tmp: String = ""

        fun createCancelTask(): TimerTask {
            return object : TimerTask() {
                override fun run() {
                    count++
                    Log.d("count", count.toString())

                    //現在沒錄音 => 啟動錄音
                    if (!isRecording) {
                        waveRecorder.startRecording()
                    }
                    //正在錄音 => 關閉錄音
                    else {
                        waveRecorder.stopRecording()
                        //存檔
                        //filepath 從 1.wav 改為 2.wav
                        if (!isWav2) {
                            filePath = recordFilePath + "/audioFile2.wav"
                            waveRecorder.changeFilePath(filePath)
                            Log.d("changefilepath",filePath )
                            isWav2 = true
                            filepath_tmp = recordFilePath + "/audioFile1.wav"
                        }
                        //filepath 從 2.wav 改為 1.wav
                        else {
                            filePath = recordFilePath + "/audioFile1.wav"
                            waveRecorder.changeFilePath(filePath)
                            Log.d("changefilepath",filePath )
                            isWav2 = false
                            filepath_tmp = recordFilePath + "/audioFile2.wav"
                        }
                        waveRecorder.startRecording()
                        transcript = SocketClient(serverAddress, serverPort, filepath_tmp).connect(filepath_tmp)
                    }

                    //讀檔生成=====================================================================
                    //val json = loadJSONFromAsset(baseContext,"chats1.json")
                    val chatList = parseChatJSON(transcript, language)
                    val currentChatList = dataSource.getChatList().value?.toMutableList() ?: mutableListOf()
                    //驗證 append/ concat
                    Log.d("chatList", chatList.toString())
                    Log.d("currentChatList", currentChatList.toString())
                    if ((currentChatList.size>0) and (chatList.isNotEmpty()))  {
                        currentChatList.removeAt(currentChatList.size - 1)
                    }
                    if(chatList.isNotEmpty()){
                        currentChatList.addAll(chatList)
                    }
                    //更新資料與畫面
                    dataSource.getChatList().postValue(currentChatList)
                    Log.d("currentChatList", currentChatList.toString())
                    runOnUiThread {
                        adapter.updateData(currentChatList)
                    }
                    recyclerView.postDelayed({
                        if(currentChatList.size>0)
                        {
                            recyclerView.smoothScrollToPosition(currentChatList.size - 1)
                        }
                    }, 100)

                }
            }
        }


        //設定按鈕=======================================================================
        val recButton = findViewById<Button>(R.id.rec_button)
        recButton.setOnClickListener {
            //尚未 create timer => set timer
            if (!isTimerScheduled) {
                recButton.setBackgroundResource(R.drawable.stop_button)
                timer = Timer()
                val cancelTask1 = createCancelTask()
                timer?.schedule(cancelTask1, 0, 4500)
                Log.d("schedule", "schedule")
                isTimerScheduled = true
            }
            //已 create timer => delete timer
            else {
                recButton.setBackgroundResource(R.drawable.record_button)
                timer?.cancel()
                Log.d("cancel", "cancel")
                isTimerScheduled = false
                waveRecorder.stopRecording()
            }
         }
    }

    //錄音系列 function==========================================================================
    private fun startRecording() {
        Log.d(TAG, waveRecorder.audioSessionId.toString())
        isRecording = true
    }

    private fun stopRecording() {
        isRecording = false
        runOnUiThread {
            //Toast.makeText(this, "File saved at : $filePath", Toast.LENGTH_SHORT).show()
        }
    }


    //權限系列 function==========================================================================
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>, grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            PERMISSIONS_REQUEST_RECORD_AUDIO -> {
                if ((grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED)) {
                    waveRecorder.startRecording()
                }
                return
            }

            else -> {
            }
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
    override fun onDestroy() {
        super.onDestroy()
        // Ensure you cancel the timer when the activity is destroyed to avoid memory leaks
        timer?.cancel()
    }
}